### ISSUE TITLE
Need a more solid approach to handle exceptions

### ISSUE THREAD

**Issue by MartinNowak** [rejectedsoftware/vibe.d#979]
- Exceptions should not be used to handle programming errors.
- There should be a configurable 5XX renderer.
- For debug builds stacktraces are interesting, for release builds at maximum the exception message should be visible.

**Comment by Geod24** [comment]
> Exceptions should not be used to handle programming errors.

Why ?
Agree with 2 and 3.

**Comment by s-ludwig** [comment]
There is `HTTPServerSettings.errorPageHandler` to customize error pages, or do you mean something else? There is also `HTTPServerOption.errorStackTraces` which is set by default, but can be disabled.

I agree that exceptions should generally not be used to detect programming errors (in contrast to assertions), but the line between programming and input error is hard to draw in many cases, so there may indeed be some borderline cases\*. Do you have one or more specific examples in mind?

\* There is one place where I'd like to make the opposite change. `BsonObjectID.fromHexString` currently uses an assertion for invalid input strings, which would require the caller to manually validate the string up front. I think this would better be handled with an exception.

**Comment by MartinNowak** [comment]
> There is HTTPServerSettings.errorPageHandler to customize error pages, or do you mean something else?

No, that's what I mean. So maybe the default should be more like point 3.
I filed this because I see stack traces on many vibe.d site for internal errors as well as for invalid input.
So clearly the defaults don't work very well here.

> There is one place where I'd like to make the opposite change. BsonObjectID.fromHexString currently uses an assertion for invalid input strings, which would require the caller to manually validate the string up front. I think this would better be handled with an exception.

Either this or use some other way to signal success, e.g. `bool parseHexString(ref BsonObjectID, string s)`.

**Comment by jacob-carlborg** [comment]
What's the status of this? I agree with point three, in non-debug mode exceptions should be written to a log file and a generic error page (without stacktrace) should be render to the client.
